### ISSUE TITLE
Add "n/m characters remaining" character counter to the appeal reason

### ISSUE THREAD

**Issue by luni3359** [danbooru/danbooru#4519]
The site only warns you after you have sent the reason, it should let you know beforehand.
![Reason](https://cdn.discordapp.com/attachments/710192137878241372/722855795636830238/unknown.png)

**Comment by BrokenEagle** [comment]
Related to #4429.

**Comment by lllusion3469** [comment]
There's e.g. HTMLs `<textarea required maxlength="140">` / `<input required maxlength="140">`
(`minlength` only applies if not empty)

This could be combined with the [`:valid`](https://developer.mozilla.org/en-US/docs/Web/CSS/:valid) / `:not(:valid)` pseudo-class and maybe [`::after`](https://developer.mozilla.org/en-US/docs/Web/CSS/::after) if you want to show text, though this wouldn't show the current length (which was requested).

One problem, though, is that HTML/JS count in "UTF-16 code units" whereas ruby counts in Unicode code points, so if characters outside the [BMP](https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane) are used it might get rejected even if it's valid in ruby. If you want the number of code points in JS, you have to do it [like this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length#Unicode)

**Comment by evazion** [comment]
HTML5 client-side validations like `maxlength` are preferable to writing our own Javascript widgets. See https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation.

SimpleForm is supposed to generate client side validations by default, but we had to disable it before because a bug in SimpleForm caused it to generate wrong `minlength` validations on everything (https://github.com/heartcombo/simple_form/issues/1496). I think that bug has been fixed, so we could probably enable it again.

As for calculating string length, technically we should count grapheme clusters rather than bytes or code points. See https://hsivonen.fi/string-length/:

```
# ruby
"🤦🏼‍♂️".grapheme_clusters.size # => 1
"🤦🏼‍♂️".codepoints.size # => 5
"🤦🏼‍♂️".bytes.size # => 17
"🤦🏼‍♂️".size # => 5

# javascript
"🤦🏼‍♂️".length # => 7
[..."🤦🏼‍♂️"].length # => 5
```

I don't think anyone - browsers, Javascript, or Rails - counts characters correctly by default.

**Comment by lllusion3469** [comment]
> As for calculating string length, technically we should count grapheme clusters rather than bytes or code points.

Yes, technically, but `maxlength` can't do that. I was only mentioning what you would have to do to get the client side (HTML/JS) to act in line with the backend (ruby).

And you'd have to do something at least a bit custom to display the current length.

**Comment by evazion** [comment]
> HTML5 client-side validations like maxlength are preferable to writing our own Javascript widgets.

On second thought, a Javascript counter might not be a bad idea. `maxlength` by itself is a little confusing, it's not obvious why you're limited once you hit the limit. Long term it would be better to introduce React or some other framework for this though. Using JQuery for UI components is too low level, we end up with too much complex Javascript.

**Comment by steve-redka** [comment]
This issue is 6 years old. Is it still worth adding backend-driven counters to fields?

**Comment by nonamethanks** [comment]
Implemented in 8eeb0ad093a8a48c8ea24f0e2d082766c4ec8ac0. Currently deployed on testbooru for testing. 

Still haven't found a good solution for composite emoji. 

<img width="1085" height="294" alt="Image" src="https://github.com/user-attachments/assets/c9805b4b-1c34-488c-835a-0adb0cd2859b" />

<img width="1050" height="262" alt="Image" src="https://github.com/user-attachments/assets/ab673761-47e8-42e1-a970-63289a8c6168" />


<img width="1478" height="323" alt="Image" src="https://github.com/user-attachments/assets/218c9aad-a58a-4217-8812-6cec8c105cfe" />

**Comment by nonamethanks** [comment]
```js
const segmenter = new Intl.Segmenter("en");
const string = "test for kanji 吾輩は猫であ 👍🏿";

console.log([...segmenter.segment(string)].length)
> 23
```

This works, but validators at the moment check for .length, so we'd have to change both JS and validators to support this.
